### ISSUE TITLE
MPD - DRM - Lost sound after two or three minutes of the live stream (HTTP 404 for the audio dash)

### ISSUE THREAD

**Issue by wwark** [peak3d/inputstream.adaptive#482]
Hi,

I work for this plugin https://github.com/Catch-up-TV-and-More/plugin.video.catchuptvandmore which using inputstream.adaptive for playing mpd content DRM protected or not.

We have HTTP 404 just for the audio while playing this mpd. (La Deux - RTBF Auvio - Belgium)
[ladeux_drm.zip](https://github.com/peak3d/inputstream.adaptive/files/4917363/ladeux_drm.zip)
During the play, the stream halts for 1 and 2 seconds (with buffering) and after that we have only video, the audio is not present.

The debug log of Kodi 19 (issue also present in kodi 18) 
[kodi-19-log.zip](https://github.com/peak3d/inputstream.adaptive/files/4917393/kodi-19-log.zip)

>2020-07-14 09:38:37.807 T:8368    DEBUG <general>: CurlFile::Open(0x1c2a327f130) https://rtbf-live.fl.freecaster.net/live/rtbf/geo/drm/ladeux_drm/dash/ladeux_drm-audio_1=128000-70326812079702.dash
2020-07-14 09:38:37.870 T:8368    ERROR <general>: AddOnLog: inputstream.adaptive: Download https://rtbf-live.fl.freecaster.net/live/rtbf/geo/drm/ladeux_drm/dash/ladeux_drm-audio_1=128000-70326812079702.dash failed with error: 404
2020-07-14 09:38:37.871 T:3760    DEBUG <general>: Stream stalled, start buffering. Audio: 0 - Video: 0

I tested the mpd and it didn't have issue.
![Screenshot from 2020-07-14 10-07-06](https://user-images.githubusercontent.com/469461/87401105-01064b00-c5ba-11ea-9dbb-a8659280120f.png)

Do not hesitate if you need us to make more tests.
Thanks,

@SylvainCecchetto and @Gaet81

**Comment by dagwieers** [comment]
It sometimes helps to pinpoint an issue by looking at how other players are handling your stream: https://github.com/add-ons/best-practices/wiki/Streaming-issues

It being DRM protected doesn't help though :disappointed:

**Comment by wwark** [comment]
we made a debugging session with @peak3d previously and it solved another case with HTTP 404 for video and audio.
So maybe it can be handle the same way for this case.

Thanks in advance !

**Comment by glennguy** [comment]
@wwark I think the best place to start would be to do side-by-side playing of the stream in Kodi and in a browser with dev tools /network open, and then compare the audio urls in the kodi log/browser. Try to compare the first segment downloaded - the browser will likely buffer ahead several segments at the start, Kodi will not. Kodi will also attempt to play ~12 seconds from the live edge, who knows what the browser will do...
The `70326812079702` part of the url is calculated dynamically by inputstream.adaptive so if it's behaving correctly in the browser then the issue must lay here.

**Comment by matthuisman** [comment]
this may be the timing issue as glenn said.

That was fixed with: https://github.com/peak3d/inputstream.adaptive/commit/98e5a72c5446afba3b7aa94aab23413787877e08
However, it's not yet in the current Leia release.

If your using Leia, try this IA build:
https://jenkins.kodi.tv/job/peak3d/job/inputstream.adaptive/job/Leia/lastSuccessfulBuild/
(zips are at the top under "Build artifacts"

That is the latest with the above commit included.

**Comment by wwark** [comment]
@glennguy and @matthuisman  thank you for your returns.

The fix https://github.com/peak3d/inputstream.adaptive/commit/98e5a72c5446afba3b7aa94aab23413787877e08 is linked to another request by me (https://github.com/peak3d/inputstream.adaptive/issues/439). I have already checked it on Kodi 19 and Kodi 18 but no luck for this stream, the audio is lost.

I will make the test this week and let you know.

**Comment by wwark** [comment]
I have another question related to the segments.
On our side we just provide mpd and drm licence if needed on Kodi/InputStream.adaptive, are there any options to force to download again the mpd or force value on segment by us ?

**Comment by glennguy** [comment]
No unfortunately, no way to force a fix from the addon's side apart from running a proxy and manually altering the mpd to correct the timing difference.
If you can test and post the different urls/segment values then we can work out the difference and it may be more obvious what's going wrong.

**Comment by ivankokan** [comment]
> this may be the timing issue as glenn said.
> 
> That was fixed with: [98e5a72](https://github.com/peak3d/inputstream.adaptive/commit/98e5a72c5446afba3b7aa94aab23413787877e08)
> However, it's not yet in the current Leia release.
> 
> If your using Leia, try this IA build:
> https://jenkins.kodi.tv/job/peak3d/job/inputstream.adaptive/job/Leia/lastSuccessfulBuild/
> (zips are at the top under "Build artifacts"
> 
> That is the latest with the above commit included.

@matthuisman So, the fix is not yet available in Kodi 18.7 Add-on repository?

**Comment by matthuisman** [comment]
You can see all the commits here:
https://github.com/peak3d/inputstream.adaptive/commits/Leia

If there is no "version bump" after a commit - then it is yet to be released to the repos.
Also, with Android - new versions can only be shipped inside the kodi build.

**Comment by ivankokan** [comment]
> You can see all the commits here:
> https://github.com/peak3d/inputstream.adaptive/commits/Leia
> 
> If there is no "version bump" after a commit - then it is yet to be released to the repos.
> Also, with Android - new versions can only be shipped inside the kodi build.

I understand: fix will be available in >2.4.5, which will for Android be available only within new Kodi build.

FYI: https://forum.kodi.tv/showthread.php?tid=124716&pid=2965914#pid2965914

**Comment by ivankokan** [comment]
Can someone please check whether the error within this log is related to the Issue (using the latest unofficial version of inputstream.adaptive @matthuisman referred to)?
https://paste.kodi.tv/idubuxatiy.kodi

**Comment by ivankokan** [comment]
Hi everyone, it turned out it was my fault: https://github.com/ivankokan/xbmc.plugin.video.nba/commit/0e04227fdb21d6233024c1e3c78510132e68793a

I had such line in previous versions. 😕 

May I ask someone to explain things better here https://github.com/peak3d/inputstream.adaptive/wiki/Integration#inputstreamadaptivemanifest_update_parameter?

**Comment by sy6sy2** [comment]
Hi,
Thank @ivankokan, it seems that the line `item.setProperty('inputstream.adaptive.manifest_update_parameter', 'full')` also fix the audio issue for us (the one described by @wwark in first message).

@peak3d, can you explain to us why this "magic" property fix our issue? Thank you very munch.

**Comment by michaelarnauts** [comment]
That property instructs Inputstream to refetch the manifest after every segment that is played. This is needed since for live content, the initial manifest contains only the last few segments up to the live point. After playing it, it needs to play new segments (that just went live), and to know these, it needs to fetch the manifest again.

You probably only need this for live content, since for vod, all the segments are already available when the stream starts.

There are some special cases related to this when the live TV service needs to have an authentication this is kept up to date. (So you need to refresh your token that you use to fetch the manifest). This is not always needed.

**Comment by matthuisman** [comment]
so it should be safe to always set it if Live content?
It doesn't cause any issues? Except more requests I guess

**Comment by michaelarnauts** [comment]
> so it should be safe to always set it if Live content?
> It doesn't cause any issues? Except more requests I guess

That's what I had to do. Otherwise, playback would stop after one segment. I think this is also what web players do, I don't see how they can know the next segments without redownloading the manifest.

I do remember an issue here however that we could just autodetect this, I'll see if I can dig up this issue.

Edit: Found it here: https://github.com/peak3d/inputstream.adaptive/pull/298#issuecomment-524206935
